### PR TITLE
Model enhancements

### PR DESCRIPTION
**Pull request by wgifford**
PatchTST fixes:
- Adjust the estimate of the mean from the forecasted quantiles.
- Use bool values for masking tensors.
- Add test cases for PatchTST with expanding context.

FlowState fixes:
- Accept `pad_mask`. Initial padded values are removed from `past_values` before existing processing.

General improvements:
- Update dataset to create a `pad_mask` to indicate where padding has occurred.
- Batches will be padded up to `max_context_length`, so `pad_mask` enables determination of padded values for each row.

### CODE REVIEW

File: tsfm_public/models/patchtst_fm/modeling_patchtst_fm.py [hunk 1]
```diff
@@ -315,21 +317,40 @@ def forward(
...
+        # quant_prob = quant_prob.view(1, -1, 1, 1).to(self.device)
+        # point_forecast: torch.Tensor = (forecast_samples * quant_prob).sum(dim=1)
 
         if not list_input:
-            quant_prob = quant_prob.view(1, -1, 1, 1).to(self.device)
-            point_forecast: torch.Tensor = (forecast_samples * quant_prob).sum(dim=1)
+            point_forecast: torch.Tensor = dq * (
+                0.5 * (forecast_samples[:, 0, ...] + forecast_samples[:, -1, ...])
+                + forecast_samples[:, 1:-1, ...].sum(dim=1)
+            )
```
Comment (hoffmansc) [inline tsfm_public/models/patchtst_fm/modeling_patchtst_fm.py]:
```suggestion
            point_forecast: torch.Tensor = torch.trapz(forecast_samples, torch.tensor(self.config.quantile_levels), dim=1)
            # extend tails
            point_forecast += forecast_samples[:, 0]*self.config.quantile_levels[0] + forecast_samples[:, -1])*self.config.quantile_levels[-1]
```

File: tsfm_public/models/patchtst_fm/modeling_patchtst_fm.py [hunk 2]
```diff
@@ -315,21 +317,40 @@ def forward(
...
         else:
-            quant_prob = quant_prob.view(-1, 1, 1).to(self.device)
-            point_forecast: List[torch.Tensor] = [(sample * quant_prob).sum(dim=0) for sample in forecast_samples]
+            point_forecast: List[torch.Tensor] = [
+                dq * (0.5 * (sample[0, ...] + sample[-1, ...]) + sample.sum(dim=0)) for sample in forecast_samples
+            ]
```
Comment (hoffmansc) [inline tsfm_public/models/patchtst_fm/modeling_patchtst_fm.py]:
```suggestion
            point_forecast: List[torch.Tensor] = [
                torch.trapz(samples, torch.tensor(self.config.quantile_levels), dim=0) +
                sample[0]*self.config.quantile_levels[0] + sample[-1]*self.config.quantile_levels[-1]
                for sample in forecast_samples
            ]
```

**Review by hoffmansc — CHANGES_REQUESTED** [review]

**Review by hoffmansc — APPROVED** [review]
